estransport/connection.go

```diff
@@ -0,0 +1,280 @@
+//
+type Connection struct {
+	sync.Mutex
+
+	URL       *url.URL
+	Dead      bool
+	DeadSince time.Time
+	Failures  int
+}
+
+type singleConnectionPool struct {
+	connection *Connection
+
+	metrics       *metrics
+	enableMetrics bool
+
+	debugLogger DebuggingLogger
+}
+
+type roundRobinConnectionPool struct {
+	sync.Mutex
+
+	curr int           // Index of the current connection
+	live []*Connection // List of live connections
+	dead []*Connection // List of dead connections
+	orig []*url.URL    // List of original URLs, passed in during initialization
+
+	metrics       *metrics
+	enableMetrics bool
+
+	debugLogger DebuggingLogger
+}
+
+func newSingleConnectionPool(u *url.URL) *singleConnectionPool {
+	cp := singleConnectionPool{connection: &Connection{URL: u}}
+
+	return &cp
+}
+
+func newRoundRobinConnectionPool(u ...*url.URL) *roundRobinConnectionPool {
+	var conns []*Connection
+	for _, url := range u {
+		conns = append(conns, &Connection{URL: url})
+	}
+
+	cp := roundRobinConnectionPool{live: conns, orig: u, curr: -1}
+
+	if cp.enableMetrics {
+		cp.metrics.Lock()
+		cp.metrics.live = cp.live
+		cp.metrics.dead = cp.dead
+		cp.metrics.Unlock()
+	}
+
+	return &cp
+}
+
+// Next returns the connection from pool.
+//
+func (cp *singleConnectionPool) Next() (*Connection, error) {
+	return cp.connection, nil
+}
+
+// Remove is a no-op for single connection pool.
+//
+func (cp *singleConnectionPool) Remove(c *Connection) error {
+	return nil
+}
+
+// Next returns a connection from pool, or an error.
+//
+func (cp *roundRobinConnectionPool) Next() (*Connection, error) {
+	var c *Connection
+
+	cp.Lock()
+	defer cp.Unlock()
+
+	// Return next live connection
+	if len(cp.live) > 0 {
+		cp.curr = (cp.curr + 1) % len(cp.live)
+		return cp.live[cp.curr], nil
+
+	} else if len(cp.dead) > 0 {
+		// Try to resurrect a dead connection if no live connections are available
+		if cp.debugLogger != nil {
+			cp.debugLogger.Log("Resurrecting connection ")
+		}
+		c, cp.dead = cp.dead[len(cp.dead)-1], cp.dead[:len(cp.dead)-1] // Pop item
+		c.Lock()
+		if cp.debugLogger != nil {
+			cp.debugLogger.Log(c.URL.String() + "\n")
+		}
+		c.markAsLive()
+		cp.live = append(cp.live, c)
+		c.Unlock()
+
+		return c, nil
+	}
+
+	return nil, errors.New("no connection available")
+}
+
+// Remove removes a connection from the pool.
+//
+func (cp *roundRobinConnectionPool) Remove(c *Connection) error {
+	cp.Lock()
+	defer cp.Unlock()
+
+	c.Lock()
+
+	if c.Dead {
+		if cp.debugLogger != nil {
+			cp.debugLogger.Logf("Already removed %s\n", c.URL)
+		}
+		c.Unlock()
+		return nil
+	}
+
+	if cp.debugLogger != nil {
+		cp.debugLogger.Logf("Removing %s...\n", c.URL)
+	}
+	c.markAsDead()
+	cp.scheduleResurrect(c)
+	c.Unlock()
+
+	// Push item to dead list and sort slice by number of failures
+	cp.dead = append(cp.dead, c)
+	sort.Slice(cp.dead, func(i, j int) bool {
+		c1 := cp.dead[i]
+		c2 := cp.dead[j]
+		c1.Lock()
+		c2.Lock()
+		defer c1.Unlock()
+		defer c2.Unlock()
```

Comment:
cp.dead can not be touched because of `cp.Lock`. Do we need to lock the mutexes here? Are connections in the pool supposed to be used by many go-routines concurrently?

Reply:
Yes, it's entirely possible that the user creates a goroutine per API call, and uses the client concurrently.

The `container/heap` package would be useful here, but I'm not sure if it would simplify things or confuse them... Will investigate at some later point.